---
title: "DE_a_p"
author: "ManasyanAL"
date: "14 08 2023"
output: html_document
---

```{r echo = T, results = 'hide'}
library(DESeq2)
library(ggplot2)
library(RColorBrewer)
library(pheatmap)
library(EnhancedVolcano)
library(ape)
library(dplyr)
library(apeglm)
```
## Loading dataset
```{r}
Counts_chd1 <- read.delim("data/chd1_featurecounts.csv", header = TRUE, row.names = 1, sep = ",")
```

## Remove genes with null expression
```{r}
Counts_chd1 <- Counts_chd1[which(rowSums(Counts_chd1) > 10),]
```

```{r}
head(Counts_chd1, 10)
dim(Counts_chd1)
```

## Creating DESeqDataSet Objects
```{r}
dds_chd1 <- DESeqDataSetFromMatrix(countData = Counts_chd1,
                                   colData = data.frame(condition = factor(c("Ctrl", "Auxin", "Ctrl", "Auxin"))),
                                   design = ~1+condition)
```

## DESeq2
```{r echo = 'hide', results = 'hide'}
dds_chd1 <- DESeq(dds_chd1)
```

```{r}
contrast_chd1 <- c("condition", "Auxin", "Ctrl")
```

## Test contrasts and get results
```{r}
res_chd1 <- results(dds_chd1, contrast =  c(contrast_chd1))
```

## Filter results by adjusting for pudge value and log2foldchange
```{r}
padj.cutoff <- 0.01
lfc.cutoff <- 0.58
```

```{r}
sig_res_chd1 <- res_chd1[which(res_chd1$padj < padj.cutoff & abs(res_chd1$log2FoldChange) > lfc.cutoff), ]
```

```{r}
summary(sig_res_chd1)
```

## Create graphs

### MAplot
```{r}
chd1_resLFC <- lfcShrink(dds_chd1, coef = "condition_Ctrl_vs_Auxin", type = "apeglm",
                    lfcThreshold = 1)
```

```{r}
plotMA(chd1_resLFC, alpha = 0.01, ylim = c(-7, 7), main = "MA Plot for Chd1")
```

### Table output with a difference in DE
```{r}
diff_expr_chd1 <- data.frame(row.names = rownames(sig_res_chd1), log2FoldChange = sig_res_chd1$log2FoldChange, padj = sig_res_chd1$padj)
```

```{r}
sorted_diff_expr_chd1 <- diff_expr_chd1[order(diff_expr_chd1$log2FoldChange), ]
head(sorted_diff_expr_chd1, 10)
```

```{r}
summary_table_chd1 <- data.frame(
  Condition = c("Chd1"),
  Up = c(sum(sig_res_chd1$log2FoldChange > lfc.cutoff)),
  Down = c(sum(sig_res_chd1$log2FoldChange < -lfc.cutoff))
)
```


```{r}
summary_table_chd1
```


# Annotation

```{r}
annotation_file <- "data/dmel-all-r6.52.gtf.gz"
annotation <- read.gff(annotation_file, na.strings = c(".", "?"), GFF3 = FALSE)
```

```{r}
annotation_filter <- annotation %>%
  filter(feature == "gene") %>%
  dplyr::select(c(1, 3, 9)) %>% 
  setNames(c("chr", "feature", "attr")) %>% 
  mutate(id = sub('.*gene_id "([^"]+)".*', '\\1', attr), chr = paste0("chr", chr)) %>% 
  dplyr::select(-attr)
```

```{r}
head(annotation_filter, 10)
dim(annotation_filter)
```

```{r}
Counts_chd1.df <- as.data.frame(sorted_diff_expr_chd1) %>% mutate(id = rownames(sorted_diff_expr_chd1))
```


```{r}
Counts_chd1.df <- merge(select(annotation_filter, id, chr), Counts_chd1.df, by = "id", all.y = T)

head(Counts_chd1.df, 10)
```

```{r}
unique(Counts_chd1.df$chr)
```

```{r}
Counts_chd1.df$group <- ifelse(Counts_chd1.df$chr == "chrX", "chrX", "Autosome")

# Рисуем боксплоты
ggplot(Counts_chd1.df, aes(x = group, y = log2FoldChange, fill = group)) +
  geom_boxplot() +
  labs(title = "Boxplot of log2FoldChange for Chd1", x = "Group", y = "log2FoldChange") +
  scale_fill_discrete(name = "Group") +
  theme_minimal() +
  ylim(-7, 5)
```

```{r}
center <- 0
spread <- 1.5
```

```{r}
ggplot(Counts_chd1.df, aes(x = log2FoldChange, fill = group, color = group)) +
  geom_density(alpha = 0.5) +
  labs(title = "Density Plot of log2FoldChange for Chd1", x = "log2FoldChange", y = "Density") +
  scale_fill_discrete(name = "Group") +
  scale_color_discrete(name = "Group") +
  theme_minimal() +
  xlim(center - 3 * spread, center + 3 * spread)
```

```{r}
# Загружаем пакет для статистического анализа
library(stats)

# Выполняем t-тест
t_test_result <- t.test(log2FoldChange ~ group, data = Counts_chd1.df)

# Выводим результаты t-теста
print("T-Test:")
print(t_test_result)

# Выполняем ранговый тест Манна-Уитни
wilcox_test_result <- wilcox.test(log2FoldChange ~ group, data = Counts_chd1.df)

# Выводим результаты рангового теста Манна-Уитни
print("Mann-Whitney U Test:")
print(wilcox_test_result)
```




